module/component/component.go

```diff
@@ -0,0 +1,309 @@
+// This is where a component should perform any necessary startup
+// tasks before worker routines are launched.
+type ComponentStartup func(context.Context) error
+
+// ComponentManagerBuilder provides a mechanism for building a ComponentManager
+type ComponentManagerBuilder interface {
+	// OnStart sets the startup routine for the ComponentManager. If an error is
+	// encountered during the startup routine, the ComponentManager will shutdown
+	// immediately.
+	OnStart(ComponentStartup) ComponentManagerBuilder
+
+	// AddWorker adds a worker routine for the ComponentManager
+	AddWorker(ComponentWorker) ComponentManagerBuilder
+
+	// AddComponent adds a new sub-component for the ComponentManager.
+	// This should be used for critical sub-components whose failure would be
+	// considered irrecoverable. For non-critical sub-components, consider using
+	// RunComponent instead.
+	AddComponent(Component) ComponentManagerBuilder
+
+	// Build builds and returns a new ComponentManager instance
+	Build() *ComponentManager
+}
+
+type ComponentManagerBuilderImpl struct {
+	startup    ComponentStartup
+	workers    []ComponentWorker
+	components []Component
+}
+
+// NewComponentManagerBuilder returns a new ComponentManagerBuilder
+func NewComponentManagerBuilder() ComponentManagerBuilder {
+	return &ComponentManagerBuilderImpl{
+		startup: func(ctx context.Context) error {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				return nil
+			}
+		},
+	}
+}
+
+func (c *ComponentManagerBuilderImpl) OnStart(startup ComponentStartup) ComponentManagerBuilder {
+	c.startup = startup
+	return c
+}
+
+func (c *ComponentManagerBuilderImpl) AddWorker(worker ComponentWorker) ComponentManagerBuilder {
+	c.workers = append(c.workers, worker)
+	return c
+}
+
+func (c *ComponentManagerBuilderImpl) AddComponent(component Component) ComponentManagerBuilder {
+	c.components = append(c.components, component)
+	return c
+}
+
+func (c *ComponentManagerBuilderImpl) Build() *ComponentManager {
+	return &ComponentManager{
+		started:     atomic.NewBool(false),
+		startupDone: make(chan struct{}),
+		ready:       make(chan struct{}),
+		startup:     c.startup,
+		workers:     c.workers,
+		components:  c.components,
+	}
+}
+
+var _ Component = (*ComponentManager)(nil)
+
+// ComponentManager is used to manage worker routines of a Component
+type ComponentManager struct {
+	started        *atomic.Bool
+	startupDone    chan struct{}
+	ready          chan struct{}
+	done           sync.WaitGroup
+	shutdownSignal <-chan struct{}
+
+	startup    func(context.Context) error
+	workers    []ComponentWorker
+	components []Component
+}
+
+// Start initiates the ComponentManager. It will first run the startup routine if one
+// was set, and then start all sub-components and launch all worker routines.
+func (c *ComponentManager) Start(parent irrecoverable.SignalerContext) (err error) {
+	if c.started.CAS(false, true) {
```

Comment:
can you add a comment about what this is doing? I had to look it up. even something simple like "Only start once"